bluetooth/src/androidLibMain/kotlin/BluetoothMonitor.kt

```diff
@@ -17,61 +17,131 @@
 
 package com.splendo.kaluga.bluetooth
 
+import android.Manifest
 import android.bluetooth.BluetoothAdapter
 import android.content.BroadcastReceiver
 import android.content.Context
 import android.content.Intent
 import android.content.IntentFilter
+import android.content.pm.PackageManager
+import android.location.LocationManager
+import androidx.core.content.ContextCompat
 import com.splendo.kaluga.base.ApplicationHolder
 import com.splendo.kaluga.base.DefaultServiceMonitor
 import com.splendo.kaluga.base.ServiceMonitor
+import com.splendo.kaluga.base.monitor.ServiceMonitorState
+import com.splendo.kaluga.logging.debug
+import kotlin.coroutines.CoroutineContext
+
+internal typealias SystemServiceState = Int
 
 actual interface BluetoothMonitor : ServiceMonitor {
 
+    /**
+     * Builder for [BluetoothMonitor]. When is used from simulator, [LocationManager] is returned as **null**.
+     * In that case the state will be emitted as [ServiceMonitorState.NotSupported] and no [BroadcastReceiver] will be registered.
+     * @param context [Context] used to register the [BroadcastReceiver].
+     * @param adapter [BluetoothAdapter] in order to define whether the service is active or not, or if service is not supported at all.
+     */
     actual class Builder(
         private val context: Context = ApplicationHolder.applicationContext,
-        private val adapter: BluetoothAdapter
+        private val adapter: BluetoothAdapter?
     ) {
-        actual fun create(): BluetoothMonitor {
+        /**
+         * Builder's create method.
+         * @param coroutineContext [CoroutineContext] used to define the coroutine context where code will run.
+         * @return [DefaultServiceMonitor]
+         */
+        actual fun create(coroutineContext: CoroutineContext): DefaultServiceMonitor {
             return DefaultBluetoothMonitor(
                 applicationContext = context,
-                bluetoothAdapter = adapter
+                bluetoothAdapter = adapter,
+                coroutineContext = coroutineContext
             )
         }
     }
 }
 
 class DefaultBluetoothMonitor internal constructor(
     private val applicationContext: Context,
-    private val bluetoothAdapter: BluetoothAdapter
-) : DefaultServiceMonitor(), BluetoothMonitor {
+    private val bluetoothAdapter: BluetoothAdapter?,
+    coroutineContext: CoroutineContext
+) : DefaultServiceMonitor(coroutineContext), BluetoothMonitor {
+
+    private val isUnauthorized: Boolean
+        get() = ContextCompat.checkSelfPermission(
+            applicationContext,
+            Manifest.permission.BLUETOOTH
+        ) == PackageManager.PERMISSION_DENIED || ContextCompat.checkSelfPermission(
+            applicationContext,
+            Manifest.permission.ACCESS_COARSE_LOCATION
+        ) == PackageManager.PERMISSION_DENIED
```

Comment:
iOS bluetooth monitor returns a state called `CBCentralManagerStateUnauthorized` when Service's permission is not granted. Therefore I added another state to `ServiceMonitorState` and adjusted code.

`Manifest.permission.BLUETOOTH` has a [**normal** protection level](https://developer.android.com/reference/android/Manifest.permission#BLUETOOTH) so `checkSelfPermission` will always returns `PackageManager.PERMISSION_GRANTED`. 
Is also true that android then in order scan for nearby devices uses location (you can replicate by requesting Kaluga`Permission.Bluetooth` and android system alert will appear asking for location. [More context](https://android.stackexchange.com/questions/160479/why-do-i-need-to-turn-on-location-services-to-pair-with-a-bluetooth-device)), so actually this property will need to check whether location was granted or not. @ChristoferAlexander @posytive @Daeda88 @vpodlesnyak what do you think about it?

Reply:
@corrado4eyes ⬆️ I think that is not done yet?

Reply:
accf0df